### PR TITLE
Fix sync tool bugs

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

The sync tool wasn't working as expected

## Description

This fixes various bugs in the sync tool and adds new features

## Testing

Tested by running the tool and running the tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tools/smithy-rs-sync/src/main.rs [hunk 1]
```diff
@@ -203,10 +234,16 @@ fn get_last_synced_commit(repo_path: &Path) -> Result<Oid> {
 /// Write the last synced commit to the file in aws-sdk-rust that tracks the last smithy-rs commit it was synced with.
 fn set_last_synced_commit(repo: &Repository, oid: &Oid) -> Result<()> {
     let repo_path = repo.workdir().expect("this will always exist");
+    let oid_string = oid.to_string();
+    let oid_bytes = oid_string.as_bytes();
     let path = repo_path.join(COMMIT_HASH_FILENAME);
-    let mut file = OpenOptions::new().write(true).truncate(true).open(&path)?;
 
-    file.write(oid.to_string().as_bytes())
+    OpenOptions::new()
```
Comment (rcoh) [inline tools/smithy-rs-sync/src/main.rs:241]: Just use std::fs::write_string?

File: tools/smithy-rs-sync/src/fs.rs [hunk 2]
```diff
@@ -135,30 +136,31 @@ impl HandwrittenFiles {
...
-    ) -> Result<impl Iterator<Item = PathBuf> + 'a, HandwrittenFilesError> {
+    pub fn handwritten_files_and_folders_iter(
+        &'_ self,
+    ) -> Result<impl Iterator<Item = PathBuf> + '_, HandwrittenFilesError> {
         self.files_and_folders_iter(FileKind::Handwritten)
     }
 
-    fn files_and_folders_iter<'a>(
-        &'a self,
+    fn files_and_folders_iter(
```
Comment (rcoh) [inline tools/smithy-rs-sync/src/fs.rs:155]: Pretty sure you can remove the first _ in these (the one on self)

**Review by rcoh — APPROVED** [review]
Lgtm!

File: tools/smithy-rs-sync/src/main.rs (hunk 1 above)
Comment (Velfi) [reply tools/smithy-rs-sync/src/main.rs:241]: good call

**Pull request merged**